latch_cli/services/cp/utils.py

```diff
@@ -0,0 +1,199 @@
+def normalize_path(path: str) -> str:
+    if legacy_expr.match(path):
+        return path  # let nuke-data deal with legacy paths
+
+    match = scheme_expr.match(path)
+    if match is None:
+        raise PathResolutionError(f"{path} is not in a valid format")
+
+    if match["missing_latch"] is not None:
+        path = f"latch{path}"
+    elif match["missing_scheme_with_leading_slash"] is not None:
+        path = f"latch://{path}"
+    elif match["missing_scheme_without_leading_slash"] is not None:
+        path = f"latch:///{path}"
+
+    from latch_cli.config.user import user_config
+
+    workspace = user_config.workspace
+
+    parsed = urlparse(path)
+    domain = parsed.netloc
+
+    if domain == "" and workspace != "":
+        domain = f"{workspace}.account"
+
+    match = domain_expr.match(domain)
+    if match is None:
+        raise PathResolutionError(f"{domain} is not a valid path domain")
+
+    if match["shared_without_selector"] is not None and workspace != "":
+        domain = f"shared.{workspace}.account"
+
+    return parsed._replace(netloc=domain).geturl()
+
+
+class LDataNodeType(str, Enum):
+    account_root = "account_root"
+    dir = "dir"
+    obj = "obj"
+    mount = "mount"
+    link = "link"
+
+
+class FinalLinkTargetPayload(TypedDict):
+    type: str
+    name: str
+
+
+class LdataNodePayload(TypedDict):
+    finalLinkTarget: FinalLinkTargetPayload
+
+
+class LdataResolvePathToNodePayload(TypedDict):
+    path: str
+    ldataNode: LdataNodePayload
+
+
+@dataclass(frozen=True)
+class GetNodeDataResult:
+    name: str
+    type: LDataNodeType
+    is_parent: bool
+
+
+def get_node_data(
+    remote_path: str, *, allow_resolve_to_parent: bool = False
+) -> GetNodeDataResult:
+    from latch.gql._execute import execute
+
+    res: LdataResolvePathToNodePayload = execute(
+        gql.gql("""
+        query GetNodeType($path: String!) {
+            ldataResolvePathToNode(path: $path) {
+                path
+                ldataNode {
+                    finalLinkTarget {
+                        name
+                        type
+                    }
+                }
+            }
+        }
+    """),
+        {"path": remote_path},
+    )["ldataResolvePathToNode"]
+
+    try:
+        final_link_target = res["ldataNode"]["finalLinkTarget"]
+        remaining = res["path"]
+
+        if not allow_resolve_to_parent and remaining is not None and remaining != "":
+            raise ValueError("Node cannot be resolved in this workspace")
+
+        is_parent = remaining is not None and remaining != ""
+
+        if remaining is not None and "/" in remaining:
+            raise ValueError("Node cannot be resolved in this workspace")
+
+        return GetNodeDataResult(
+            name=final_link_target["name"],
+            type=LDataNodeType(final_link_target["type"].lower()),
+            is_parent=is_parent,
+        )
+    except (TypeError, ValueError) as e:
+        raise PathResolutionError(
```

Comment:
this error message is not great:
- no color
- "this workspace" - which?
- `remote_path` gets lost in the message
- the instructions are a bit overwhelming, could be more clear and sympathetic

I'd do:
```
<red>
<b>{remote_path}:</b> not found # use original remote path here, not normalized

<muted>
<b>Relative path resolved to:</b> {normalized_remote_path} # iff original path was account-relative
<b>Using workspace:</b> {workspace_name} ({workspace_id}) # iff original path was account-relative

<b>Authorized using:</b> SDK token # iff using SDK auth
<b>Authorized using:</b> Execution token # iff using execution auth

<b>Check that:</b>
1. Target object exists
2. Account <b>{signer_account}</b> has permission to view the target object # signer_account is execution created_by or SDK token owner
3. The correct workspace is selected # iff original path was account-relative

For privacy reasons, non-viewable objects and non-existent objects are indistinguishable
</muted>
</red>
```